File: Emulator/Controller.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace NesEmu.Emulator
{
    public class Controller
    {
        public bool Up;
        public bool Down;
        public bool Left;
        public bool Right;

        public bool A;
        public bool B;

        public bool Select;
        public bool Start;

        private bool strobe;
        private byte noise;
        private byte state;

        public void Write(byte data)
        {
            if ((data & 0x01) == 0 && this.strobe)
            {
                CaptureState();
            }

            this.noise = (byte)(data & 0xf4);
            this.strobe = (data & 0x01) != 0;
        }

        public byte Read()
        {
            if (this.strobe)
            {
                this.CaptureState();
            }

            var status = this.state >> 7;
            this.state <<= 1;
            this.state |= 1;

            return (byte)(this.noise | status);
        }


        private void CaptureState()
        {
            this.state = 0;
            if (this.A)
                this.state |= 0x80;
            if (this.B)
                this.state |= 0x40;
            if (this.Select)
                this.state |= 0x20;
            if (this.Start)
                this.state |= 0x10;
            if (this.Up)
                this.state |= 0x08;
            if (this.Down)
                this.state |= 0x04;
            if (this.Left)
                this.state |= 0x02;
            if (this.Right)
                this.state |= 0x01;
        }
    }
}

File: NesCoreCpp/Controller.cpp
#include "Controller.h"


void Controller::ButtonDown(Buttons button)
{
    state_.Buttons |= button;
}

void Controller::ButtonUp(Buttons button)
{
    state_.Buttons &= ~button;
}

void Controller::SetButtonState(int32_t buttons)
{
    state_.Buttons = buttons;
}

uint8_t Controller::Read()
{
    if (state_.Strobe)
    {
        CaptureState();
    }

    auto status = state_.State >> 7;
    state_.State <<= 1;
    state_.State |= 1;

    return state_.Noise | status;
}

void Controller::Write(uint8_t data)
{
    if (!(data & 0x01) && state_.Strobe)
    {
        CaptureState();
    }

    state_.Noise = data & 0xf4;
    state_.Strobe = data & 0x01;
}

void Controller::CaptureState(ControllerState* state) const
{
    *state = state_;
}

void Controller::RestoreState(const ControllerState& state)
{
    state_ = state;
}

void Controller::CaptureState()
{
    uint32_t buttons = state_.Buttons;

    // Prevent inputs that are supposed to be impossible from a NES controller here - this can cause horrible
    // crashes in some games, e.g. Burger Time
    auto upAndDown = (Buttons::Up | Buttons::Down);
    if ((buttons & upAndDown) == upAndDown)
        buttons &= ~upAndDown;

    auto leftAndRight = (Buttons::Left | Buttons::Right);
    if ((buttons & leftAndRight) == leftAndRight)
        buttons &= ~leftAndRight;

    state_.State = static_cast<uint8_t>(buttons);
}
